resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/SafeFileOutputStream.java

```diff
@@ -46,71 +49,66 @@ public SafeFileOutputStream(File file) throws IOException {
 	 * @param tempPath The temporary location to use, or <code>null</code> to
 	 * use the same location as the target path but with a different extension.
 	 */
+	@SuppressWarnings("unused")
 	public SafeFileOutputStream(String targetPath, String tempPath) throws IOException {
-		failed = false;
-		target = new File(targetPath);
-		createTempFile(tempPath);
+		this.tempPath = tempPath != null ? tempPath : (targetPath + EXTENSION);
+		this.targetPath = targetPath;
+		output = new ByteArrayOutputStream();
+	}
+
+	@Override
+	public void close() throws IOException {
+		File target = new File(targetPath);
+		File temp = getTempFile();
 		if (!target.exists()) {
 			if (!temp.exists()) {
-				output = new BufferedOutputStream(new FileOutputStream(target));
+				Files.write(target.toPath(), output.toByteArray());
 				return;
 			}
 			// If we do not have a file at target location, but we do have at temp location,
-			// it probably means something wrong happened the last time we tried to write it.
+			// it probably means something wrong happened the last time we tried to write
+			// it.
 			// So, try to recover the backup file. And, if successful, write the new one.
 			Files.copy(temp.toPath(), target.toPath());
 		}
-		output = new BufferedOutputStream(new FileOutputStream(temp));
-	}
+		byte[] oldContent = Files.readAllBytes(target.toPath());
+		byte[] newContent = output.toByteArray();
```

Comment:
This loads 2 the size of the file in memory. For small files it's fine, but if file has some risk of reaching some MB, it can the be the cause of GC triggering and so on, which in the end would also be annoying.
Would it be possible to compare byte-to-byte, or to use some utility that compares stream without loading them fully in memroy?

Reply:
See the related issue: even all files together are <12 MB on a big workspace. There is no need for memory optimizations.

Reply:
@mickaelistria this a a classical task for a `RandomAccessFile` that can be opened as R/W, so you can read the bytes from the one stream, compare it to the one read by `RandomAccessFile` and then skip over if it is the same or write the bytes, this will only rewrite bytes that are changed, one must only take into account that `RandomAccessFile` is quite bare, so reading things buffered (e.g. 4096 byte buffer) will improve performance.

Reply:
@jukzi if you are hunting for *seconds* a read/compare/write can save you even more seconds, as you need to read the old state anyways, combining the compare with the write is more performant and `RandomAccessFile` is even memory mapped if you like ;-)

Reply:
> See the related issue: even all files together are <12 MB on a big workspace

Please define "big". We've seen .markers files with ~200 MB. Place them on NFS and have fun :).

Reply:
And on that workspace 200 MB won't bother you too.

Reply:
> And on that workspace 200 MB won't bother you too.

Not sure what do you mean, if I say you that it is slow to read 200 MB, and you say I shouldn't bother - what kind of argument is needed to explain that extra read of 200 MB from NFS may double the time needed to write workspace data on slow file system.

Reply:
BTW, you are trying to optimize 1 second case and ignore possible regression of **minutes**.

Reply:
The memory size complained about was java heap memory. If you are talking about file transfer time you might wonder that this PR can totally eliminate the network transfer, as the file content is normally cached by OS anyway and this PR in that case can prevents any read/write via network if the contents stays the same (which is the normal case).

Reply:
> can totally eliminate the network transfer, as the file content is normally cached by OS anyway and

**Can** and **normally** indicates that you are not sure that this will be the case and of course no one can guarantee that.

What is however guaranteed, that there will be an OS that would not have caching enabled or the cache will be too small or it will be **different** virus scanner running as one you've tested and it **will** prevent OS from caching the data because it **will** want to read / check every accessed file.

We are talking about N different operating systems with M versions, X different virus scanners with Y versions - and that only if we have data locally (client side), but of course we might have NFS server side with similar combination of various components that would be involved.

With that, I'm pretty sure that this PR will introduce a regression and therefore I'm against merging this PR "as is".